Testamos diversos modelos e selecionamos aquele que tinha o menor erro. O que mais podemos fazer?

Vamos carregar os nossos arquivos novamente.

In [0]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn.metrics import mean_absolute_error
from sklearn.ensemble import RandomForestRegressor

In [0]:
dados_treino = pd.read_csv('treino.csv', index_col='Id')
dados_teste = pd.read_csv('teste.csv', index_col='Id')

Legal, agora podemos inspecioná-los melhor. 

In [24]:
dados_treino.head(10)

MSSubClass MSZoning  LotFrontage  ...  SaleType SaleCondition SalePrice
Id                                    ...                                  
1           60       RL         65.0  ...        WD        Normal    208500
2           20       RL         80.0  ...        WD        Normal    181500
3           60       RL         68.0  ...        WD        Normal    223500
4           70       RL         60.0  ...        WD       Abnorml    140000
5           60       RL         84.0  ...        WD        Normal    250000
6           50       RL         85.0  ...        WD        Normal    143000
7           20       RL         75.0  ...        WD        Normal    307000
8           60       RL          NaN  ...        WD        Normal    200000
9           50       RM         51.0  ...        WD       Abnorml    129900
10         190       RL         50.0  ...        WD        Normal    118000

[10 rows x 80 columns]

Repare no conteúdo da coluna Alley. O que são esses NaN? NaN significa not a number e nesse caso, de que não temos um valor nessa coluna, o que não ajuda o nosso modelo, uma vez que ele se baseia em dados. 

Quantos NaN será que temos nessa coluna? Para responder essa pergunta, primeiro vemos quais deles são nulos com isnull().

In [25]:
dados_treino.Alley.isnull()

Id
1        True
2        True
3        True
4        True
5        True
6        True
7        True
8        True
9        True
10       True
11       True
12       True
13       True
14       True
15       True
16       True
17       True
18       True
19       True
20       True
21       True
22      False
23       True
24       True
25       True
26       True
27       True
28       True
29       True
30       True
        ...  
1431     True
1432     True
1433    False
1434     True
1435     True
1436     True
1437     True
1438     True
1439     True
1440     True
1441     True
1442     True
1443     True
1444     True
1445     True
1446     True
1447     True
1448     True
1449     True
1450     True
1451     True
1452     True
1453     True
1454     True
1455    False
1456     True
1457     True
1458     True
1459     True
1460     True
Name: Alley, Length: 1460, dtype: bool

Repare que isso volta um booleano para a gente e agora fazemos a soma desses valores com sum().

In [26]:
dados_treino.Alley.isnull().sum()

1369

E olhe só, de 1460 entradas, 1369 não tem valores. O que será que podemos fazer quanto a isso? Tem alguns métodos, uma saída é simplesmente jogar essa coluna fora. Para fazer isso usaríamos .drop(). Primeiro vou fazer uma cópia dos dados.

In [0]:
dados_treino_copiado = dados_treino.copy()

Passamos o axis=1 para o drop para indicar que é uma coluna.

In [28]:
dados_treino_copiado.drop('Alley', axis=1)

MSSubClass MSZoning  LotFrontage  ...  SaleType SaleCondition SalePrice
Id                                      ...                                  
1             60       RL         65.0  ...        WD        Normal    208500
2             20       RL         80.0  ...        WD        Normal    181500
3             60       RL         68.0  ...        WD        Normal    223500
4             70       RL         60.0  ...        WD       Abnorml    140000
5             60       RL         84.0  ...        WD        Normal    250000
6             50       RL         85.0  ...        WD        Normal    143000
7             20       RL         75.0  ...        WD        Normal    307000
8             60       RL          NaN  ...        WD        Normal    200000
9             50       RM         51.0  ...        WD       Abnorml    129900
10           190       RL         50.0  ...        WD        Normal    118000
11            20       RL         70.0  ...        WD        Normal    129500
12            60       RL         85.0  ...       New       Partial    345000
13            20       RL          NaN  ...        WD        Normal    144000
14            20       RL         91.0  ...       New       Partial    279500
15            20       RL          NaN  ...        WD        Normal    157000
16            45       RM         51.0  ...        WD        Normal    132000
17            20       RL          NaN  ...        WD        Normal    149000
18            90       RL         72.0  ...        WD        Normal     90000
19            20       RL         66.0  ...        WD        Normal    159000
20            20       RL         70.0  ...       COD       Abnorml    139000
21            60       RL        101.0  ...       New       Partial    325300
22            45       RM         57.0  ...        WD        Normal    139400
23            20       RL         75.0  ...        WD        Normal    230000
24           120       RM         44.0  ...        WD        Normal    129900
25            20       RL          NaN  ...        WD        Normal    154000
26            20       RL        110.0  ...        WD        Normal    256300
27            20       RL         60.0  ...        WD        Normal    134800
28            20       RL         98.0  ...        WD        Normal    306000
29            20       RL         47.0  ...        WD        Normal    207500
30            30       RM         60.0  ...        WD        Normal     68500
...          ...      ...          ...  ...       ...           ...       ...
1431          60       RL         60.0  ...        WD        Normal    192140
1432         120       RL          NaN  ...        WD        Normal    143750
1433          30       RL         60.0  ...        WD        Normal     64500
1434          60       RL         93.0  ...        WD        Normal    186500
1435          20       RL         80.0  ...        WD        Normal    160000
1436          20       RL         80.0  ...       COD       Abnorml    174000
1437          20       RL         60.0  ...        WD        Normal    120500
1438          20       RL         96.0  ...       New       Partial    394617
1439          20       RM         90.0  ...        WD        Normal    149700
1440          60       RL         80.0  ...        WD        Normal    197000
1441          70       RL         79.0  ...        WD        Normal    191000
1442         120       RM          NaN  ...        WD        Normal    149300
1443          60       FV         85.0  ...        WD        Normal    310000
1444          30       RL          NaN  ...        WD        Normal    121000
1445          20       RL         63.0  ...        WD        Normal    179600
1446          85       RL         70.0  ...        WD        Normal    129000
1447          20       RL          NaN  ...        WD        Normal    157900
1448          60       RL         80.0  ...        WD        Normal    240000
1449          50       RL   

Repare que não temos mais a nossa coluna, mas será que essa foi uma boa idéia, pois tinhamos outros valores lá que foram embora uma vez que jogamos toda a coluna fora, então uma outra abordagem seria, por exemplo, colocar outros valores nessa coluna no lugar dos NaN. Mas precisamos de valores que não tendenciariam os nossos dados, por exemplo, substituir pela media. Mas como vamos substituir pela média se o conteúdos dessas colunas são palavras?

Vamos dar um passo de cada vez. Primeiro vamos pegar apenas dados numéricos e entender como podemos fazer com eles, feito isso, vamos pegar dados em texto e ver o que podemos fazer com eles.

Como pegamos apenas os numéricos? Tudo o que é palavra é object, basta excluírmos os object e ver o que sobra para a gente. Faremos isso usando o exclude do select_dtypes()

In [29]:
dados_treino_copiado.select_dtypes(exclude=['object'])

MSSubClass  LotFrontage  LotArea  ...  MoSold  YrSold  SalePrice
Id                                      ...                           
1             60         65.0     8450  ...       2    2008     208500
2             20         80.0     9600  ...       5    2007     181500
3             60         68.0    11250  ...       9    2008     223500
4             70         60.0     9550  ...       2    2006     140000
5             60         84.0    14260  ...      12    2008     250000
6             50         85.0    14115  ...      10    2009     143000
7             20         75.0    10084  ...       8    2007     307000
8             60          NaN    10382  ...      11    2009     200000
9             50         51.0     6120  ...       4    2008     129900
10           190         50.0     7420  ...       1    2008     118000
11            20         70.0    11200  ...       2    2008     129500
12            60         85.0    11924  ...       7    2006     345000
13            20          NaN    12968  ...       9    2008     144000
14            20         91.0    10652  ...       8    2007     279500
15            20          NaN    10920  ...       5    2008     157000
16            45         51.0     6120  ...       7    2007     132000
17            20          NaN    11241  ...       3    2010     149000
18            90         72.0    10791  ...      10    2006      90000
19            20         66.0    13695  ...       6    2008     159000
20            20         70.0     7560  ...       5    2009     139000
21            60        101.0    14215  ...      11    2006     325300
22            45         57.0     7449  ...       6    2007     139400
23            20         75.0     9742  ...       9    2008     230000
24           120         44.0     4224  ...       6    2007     129900
25            20          NaN     8246  ...       5    2010     154000
26            20        110.0    14230  ...       7    2009     256300
27            20         60.0     7200  ...       5    2010     134800
28            20         98.0    11478  ...       5    2010     306000
29            20         47.0    16321  ...      12    2006     207500
30            30         60.0     6324  ...       5    2008      68500
...          ...          ...      ...  ...     ...     ...        ...
1431          60         60.0    21930  ...       7    2006     192140
1432         120          NaN     4928  ...      10    2009     143750
1433          30         60.0    10800  ...       8    2007      64500
1434          60         93.0    10261  ...       5    2008     186500
1435          20         80.0    17400  ...       5    2006     160000
1436          20         80.0     8400  ...       7    2008     174000
1437          20         60.0     9000  ...       5    2007     120500
1438          20         96.0    12444  ...      11    2008     394617
1439          20         90.0     7407  ...       4    2010     149700
1440          60         80.0    11584  ...      11    2007     197000
1441          70         79.0    11526  ...       9    2008     191000
1442         120          NaN     4426  ...       5    2008     149300
1443          60         85.0    11003  ...       4    2009     310000
1444          30          NaN     8854  ...       5    2009     121000
1445          20         63.0     8500  ...      11    2007     179600
1446          85         70.0     8400  ...       5    2007     129000
1447          20          NaN    26142  ...       4    2010     157900
1448          60         80.0    10000  ...      12    2007     240000
1449          50         70.0    11767  ...       5    2007     112000
1450         180         21.0     1533  ...       8    2006      92000
1451          90         60.0     9000  ...       9    2009     136000
1452          20         78.0     9262  ...       5    2009     287090
1453         180         35.0     3675  ...       5    2006     145000
1454          20         90.0 

Veja que agora ficamos com 37 colunas, antes tinhamos 80. 

Repare também que o SalePrice, a coluna que queremos advinhar esta no nosso X, e como querêmos prevê-la, isso não é uma boa idéia, ou seja, vamos ter que tirá-la do X também. Mas antes vamos ver se ela tem valores faltantes e atribui-la a y.

Para fazer isso, usamos o dropna, e se tiver uma linha que vai ser o axis 0 aqui que tenha Nas no SalePrice, dropamos essa linha.

In [0]:
dados_treino_copiado.dropna(axis=0, subset=['SalePrice'], inplace=True)

Agora nosso y está pronto para receber os dados de SalePrice.

In [0]:
y = dados_treino_copiado.SalePrice

Agora que já passamos o SalePrice para o y, podemos excluir a coluna (axis=1) dos dados de treino.

In [0]:
dados_treino_copiado.drop(['SalePrice'], axis=1, inplace=True)

Temos o nosso y, mas e o nosso X? Vamos fazer a atribuição e ver se está tudo certo.

In [0]:
X = dados_treino_copiado

In [34]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Data columns (total 79 columns):
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1201 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
Alley            91 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
Utilities        1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-null object
Exterior2nd      1460 non-

Repare que SalePrice não está mais no nosso X, no entanto, ele está cheio de objects, porque? Porque não fizemos in place que nem os outros métodos e não atribuimos a X, então vamos tirar os object novamente e atribuir o resultado a X. 

In [0]:
X = dados_treino_copiado.select_dtypes(exclude=['object'])

In [36]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Data columns (total 36 columns):
MSSubClass       1460 non-null int64
LotFrontage      1201 non-null float64
LotArea          1460 non-null int64
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
MasVnrArea       1452 non-null float64
BsmtFinSF1       1460 non-null int64
BsmtFinSF2       1460 non-null int64
BsmtUnfSF        1460 non-null int64
TotalBsmtSF      1460 non-null int64
1stFlrSF         1460 non-null int64
2ndFlrSF         1460 non-null int64
LowQualFinSF     1460 non-null int64
GrLivArea        1460 non-null int64
BsmtFullBath     1460 non-null int64
BsmtHalfBath     1460 non-null int64
FullBath         1460 non-null int64
HalfBath         1460 non-null int64
BedroomAbvGr     1460 non-null int64
KitchenAbvGr     1460 non-null int64
TotRmsAbvGrd     1460 non-null int64
Fireplaces       1460 non-null int64
G

Beleza, agora temos 36 colunas de inteiros e alguns floats no nosso X, e o nosso X_teste? O que fazemos com eles? 

Por enquanto não vamos alterar os valores do nosso X_teste, mas podemos ter X com 36 colunas e X_teste com 80? 

Não, certo, então para ficar justo vamos também testar nosso modelo apenas com valores numéricos.

In [0]:
X_teste = dados_teste.copy().select_dtypes(exclude=['object'])

In [39]:
X_teste.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1459 entries, 1461 to 2919
Data columns (total 36 columns):
MSSubClass       1459 non-null int64
LotFrontage      1232 non-null float64
LotArea          1459 non-null int64
OverallQual      1459 non-null int64
OverallCond      1459 non-null int64
YearBuilt        1459 non-null int64
YearRemodAdd     1459 non-null int64
MasVnrArea       1444 non-null float64
BsmtFinSF1       1458 non-null float64
BsmtFinSF2       1458 non-null float64
BsmtUnfSF        1458 non-null float64
TotalBsmtSF      1458 non-null float64
1stFlrSF         1459 non-null int64
2ndFlrSF         1459 non-null int64
LowQualFinSF     1459 non-null int64
GrLivArea        1459 non-null int64
BsmtFullBath     1457 non-null float64
BsmtHalfBath     1457 non-null float64
FullBath         1459 non-null int64
HalfBath         1459 non-null int64
BedroomAbvGr     1459 non-null int64
KitchenAbvGr     1459 non-null int64
TotRmsAbvGrd     1459 non-null int64
Fireplaces       1459 n

O que fizemos até agora? 

Nós tiramos valores faltantes das linhas do SalePrice, removemos o SalePrice de X, fizemos a atribuição do SalePrice a y, também definimos X e X_teste para ter apenas valores numéricos.

Isso resolve o nosso problema inicial de valores faltantes no dataset?

Não, vamos ver isso jajá, o que fizemos até agora foi pré processar os nossos dados de entrada.

Agora que isso está feito, vamos dividir esses dados para colocá-los no nosso modelo.

In [0]:
X_treino, X_valid, y_treino, y_valid = train_test_split(X, y,
                                                        train_size=0.8,
                                                        test_size=0.2,
                                                        random_state=42)

Beleza, agora podemos começar a mexer nos nossos dados como queríamos fazer no começo. Primeiro vamos entender quantas colunas e linhas temos nos dados de treino. Fazemos isso com .shape.

In [42]:
X_treino.shape

(1168, 36)

Repare que o shape nos devolve que são 1168 linhas e 36 colunas, e quais dessas colunas tem valores nulos?

Já sabemos como ver isso usando is null e sum.

In [0]:
colunas_com_valores_faltantes = X_treino.isnull().sum()

In [45]:
colunas_com_valores_faltantes

MSSubClass         0
LotFrontage      217
LotArea            0
OverallQual        0
OverallCond        0
YearBuilt          0
YearRemodAdd       0
MasVnrArea         6
BsmtFinSF1         0
BsmtFinSF2         0
BsmtUnfSF          0
TotalBsmtSF        0
1stFlrSF           0
2ndFlrSF           0
LowQualFinSF       0
GrLivArea          0
BsmtFullBath       0
BsmtHalfBath       0
FullBath           0
HalfBath           0
BedroomAbvGr       0
KitchenAbvGr       0
TotRmsAbvGrd       0
Fireplaces         0
GarageYrBlt       64
GarageCars         0
GarageArea         0
WoodDeckSF         0
OpenPorchSF        0
EnclosedPorch      0
3SsnPorch          0
ScreenPorch        0
PoolArea           0
MiscVal            0
MoSold             0
YrSold             0
dtype: int64

Repare que isso devolve para a gente as colunas com valores faltantes, a quantidade desses valores e também as que tem 0, eu quero ver apenas aquelas que tem valores faltantes.

Para fazer isso, podemos passar uma condição para o dataframe para que ele liste os valores maiores que 0.

In [46]:
colunas_com_valores_faltantes[colunas_com_valores_faltantes > 0]

LotFrontage    217
MasVnrArea       6
GarageYrBlt     64
dtype: int64

Já sabemos que temos colunas com valores faltantes. Agora precisamos escolher o que vamos fazer com elas, será que temos um número muito grande de valores faltantes?

Se temos 1168 entradas, e os valores faltantes são:

In [48]:
(217 + 6 + 64)/1168

0.2457191780821918

Veja que 24 a 25% dos nossos dados estão faltando, isso não é tão significativo, podemos tentar uma primeira abordagem dropando esses valores e vendo que resultado isso tem no nosso modelo.

Primeiro, vamos pegar o nome das colunas com valores faltantes.

In [0]:
colunas_com_valores_faltantes = [coluna for coluna in X_treino.columns if X_treino[coluna].isnull().any()]

O que estamos fazendo aqui é determinando que para cada coluna nas colunas de treino, se ela tiver qualquer (any) valor nulo, guardamos o nome dela na lista.

In [53]:
colunas_com_valores_faltantes

['LotFrontage', 'MasVnrArea', 'GarageYrBlt']

Beleza, agora que temos o nome das nossas colunas, podemos dropa-las do treino e da validação.

In [0]:
X_treino_reduzido = X_treino.drop(colunas_com_valores_faltantes, axis=1)
X_valid_reduzido = X_valid.drop(colunas_com_valores_faltantes, axis=1)

Agora que limpamos os dados do nosso modelo de valores faltantes, podemos criar o modelo, ajustar, predizer e calcular o erro para ver que resultado isso teve.

In [0]:
modelo = RandomForestRegressor(n_estimators=50, random_state=42)
modelo.fit(X_treino_reduzido, y_treino)
preds = modelo.predict(X_valid_reduzido)
mae = mean_absolute_error(y_valid, preds)

In [94]:
mae

17877.492979452054

Legal, parece que que o nosso erro baixou bastante.

Usamos a abordagem de dropar as colunas, e se a gente mexesse nos valores como tinhamos falado? Será que é menor ou pior?

Fazer isso é chamado de imputação, estamos imputando valores, vamos usar algo que chama Simple Imputer do sklearn.



In [0]:
from sklearn.impute import SimpleImputer

imputador_simples = SimpleImputer()

Com o imputador criado, agora vamos fazer algo que é parecido com o que fazemos com o modelo para os nossos dados. Vamos ajustar e transforma-los.

In [0]:
X_treino_imputado = pd.DataFrame(imputador_simples.fit_transform(X_treino))

In [64]:
X_treino.head(5)

MSSubClass  LotFrontage  LotArea  ...  MiscVal  MoSold  YrSold
Id                                      ...                         
255           20         70.0     8400  ...        0       6    2010
1067          60         59.0     7837  ...        0       5    2009
639           30         67.0     8777  ...        0       5    2008
800           50         60.0     7200  ...        0       6    2007
381           50         50.0     5000  ...        0       5    2010

[5 rows x 36 columns]

In [63]:
X_treino_imputado.head(5)

0     1       2    3    4       5   ...   30   31   32   33   34      35
0  20.0  70.0  8400.0  5.0  6.0  1957.0  ...  0.0  0.0  0.0  0.0  6.0  2010.0
1  60.0  59.0  7837.0  6.0  7.0  1993.0  ...  0.0  0.0  0.0  0.0  5.0  2009.0
2  30.0  67.0  8777.0  5.0  7.0  1910.0  ...  0.0  0.0  0.0  0.0  5.0  2008.0
3  50.0  60.0  7200.0  5.0  7.0  1937.0  ...  0.0  0.0  0.0  0.0  6.0  2007.0
4  50.0  50.0  5000.0  5.0  6.0  1924.0  ...  0.0  0.0  0.0  0.0  5.0  2010.0

[5 rows x 36 columns]

Repare que agora onde antes tinhamos valores faltantes agora temos uns números. Esses números são a média do valor de cada coluna, como? 

Porque apenas criamos o Simple imputer e o default dele é preencher com a média.

Repare também que perdemos a informação do nome das colunas, então vamos voltar isso.

In [0]:
X_treino_imputado.columns = X_treino.columns

In [67]:
X_treino_imputado.head(5)

MSSubClass  LotFrontage  LotArea  ...  MiscVal  MoSold  YrSold
0        20.0         70.0   8400.0  ...      0.0     6.0  2010.0
1        60.0         59.0   7837.0  ...      0.0     5.0  2009.0
2        30.0         67.0   8777.0  ...      0.0     5.0  2008.0
3        50.0         60.0   7200.0  ...      0.0     6.0  2007.0
4        50.0         50.0   5000.0  ...      0.0     5.0  2010.0

[5 rows x 36 columns]

Agora vamos fazer o mesmo para X_valid com a diferença de que não faremos o fit, apenas o transform. Do mesmo modo que fazemos com o nosso modelo, não queremos que o imputador aprenda o nosso padrão de avaliação.

In [65]:
X_valid.head(5)

MSSubClass  LotFrontage  LotArea  ...  MiscVal  MoSold  YrSold
Id                                      ...                         
893           20         70.0     8414  ...        0       2    2006
1106          60         98.0    12256  ...        0       4    2010
414           30         56.0     8960  ...        0       3    2010
523           50         50.0     5000  ...        0      10    2006
1037          20         89.0    12898  ...        0       9    2009

[5 rows x 36 columns]

In [0]:
X_valid_imputado = pd.DataFrame(imputador_simples.transform(X_valid))

In [69]:
X_valid_imputado.head(5)

0     1        2    3    4       5   ...   30   31   32   33    34      35
0  20.0  70.0   8414.0  6.0  8.0  1963.0  ...  0.0  0.0  0.0  0.0   2.0  2006.0
1  60.0  98.0  12256.0  8.0  5.0  1994.0  ...  0.0  0.0  0.0  0.0   4.0  2010.0
2  30.0  56.0   8960.0  5.0  6.0  1927.0  ...  0.0  0.0  0.0  0.0   3.0  2010.0
3  50.0  50.0   5000.0  6.0  7.0  1947.0  ...  0.0  0.0  0.0  0.0  10.0  2006.0
4  20.0  89.0  12898.0  9.0  5.0  2007.0  ...  0.0  0.0  0.0  0.0   9.0  2009.0

[5 rows x 36 columns]

Voltando as colunas.

In [0]:
X_valid_imputado.columns = X_valid.columns

Agora podemos novamente fazer o fit, predict e calcular o erro.

In [0]:
modelo_imputacao = RandomForestRegressor(n_estimators=50, random_state=42)
modelo_imputacao.fit(X_treino_imputado, y_treino)
preds_imputado = modelo_imputacao.predict(X_valid_imputado)
mae_imputado = mean_absolute_error(y_valid, preds_imputado)

In [76]:
mae_imputado

18214.297625570776

In [79]:
mae

17965.12680365297

Repare que se compararmos o mae imputado com o mae anterior, este era menor. Porque será?

Veja que ao excluirmos as colunas, perdemos bastante informação, o que diminui a complexidade do que o nosso modelo está processando e é isso que faz o erro ficar menor.

Além disso, repare que estamos usando a média dos valores do dataset para fazermos a imputação, será que essa é a melhor abordagem?

Temos alguns valores aqui numa coluna como 

60, 7, 3, 5, 1, NaN

A média desses valores seria 15, então ficaríamos com 

60, 7, 3, 5, 1, 15 

será que não estamos tendenciando um pouco os nossos dados? Repare que quando fazemos a média, estamos considerando tanto o valor 60 quanto o valor 1, e o 60 vai pesar mais.

O que poderíamos fazer então? Podemos usar algo que tendencie menos. Se ordenarmos os nossos números, temos:

1, 3, 5, 7, 60 e NaN

Pegando o do meio, ficaríamos com 5, essa é a mediana dos nossos dados e aí teríamos

1, 3, 5, 7, 60, 5

Parece uma idéia melhor do que a média, vamos ver qual o resultado disso para o nosso conjunto de dados.

Vamos repetir o mesmo processo de imputação, agora com um novo imputador cuja estratégia é usar a mediana.

In [0]:
imputador_mediana = SimpleImputer(strategy='median')

In [0]:
X_treino_mediana = pd.DataFrame(imputador_mediana.fit_transform(X_treino))
X_valid_mediana = pd.DataFrame(imputador_mediana.transform(X_valid))

In [0]:
X_treino_mediana.columns = X_treino.columns
X_valid_mediana.columns = X_valid.columns

In [0]:
modelo_mediana = RandomForestRegressor(n_estimators=50, random_state=42)
modelo_mediana.fit(X_treino_mediana, y_treino)

preds_mediana = modelo_mediana.predict(X_valid_mediana)
mae_mediana = mean_absolute_error(y_valid, preds_mediana)

In [97]:
print('MAE do modelo %s: %d' % ('com valores dropados', mae))
print('MAE do modelo %s: %d' % ('imputado com média', mae_imputado))
print('MAE do modelo %s: %d' % ('imputado com mediana', mae_mediana))

MAE do modelo com valores dropados: 17877
MAE do modelo imputado com média: 18107
MAE do modelo imputado com mediana: 18123


Como o erro da mediana está apenas um pouco maior que o da média e é uma abordagem que considera mais informações e tendencia menos os dados, vou escolher esse último modelo.

Agora vamos fazer as predições finais com os dados de teste. 

In [100]:
preds_teste = modelo_mediana.predict(X_teste)

ValueError: ignored

Repare que tivemos um erro aqui, porque? 

Lembra que deixamos o nosso X teste com colunas numéricas mas que não mexemos nos valores dele? 

Agora como ele contém NaN não conseguimos prever. Então precisamos também fazer a imputação da mediana para o X teste.

In [0]:
X_teste_mediana = pd.DataFrame(imputador_mediana.transform(X_teste))

Agora podemos fazer a predição e salvar mais um csv com os resultados.

In [0]:
preds_teste = modelo_mediana.predict(X_teste_mediana)

In [0]:
resultados_imputacao = {'Id': X_teste.index, 'SalePrice': preds_teste}

In [0]:
resultados = pd.DataFrame(resultados_imputacao)

In [0]:
resultados.to_csv('resultados_imputacao.csv', index=False)